olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,191 @@
+            download_file.write(download_stream.readall())
+
+        # model config
+        with open(model_config_path) as file:
+            model_config_dict = json.load(file)
+            return ModelConfig.from_json(model_config_dict)
+
+    def upload_model_to_cloud_cache(self, output_model_hash: str, output_model_config: ModelConfig):
+        logger.info("Uploading model to cloud cache.")
+        model_path = output_model_config.config.get("model_path")
+
+        if model_path is None:
+            logger.error("Model path is not found in the output model config. Upload failed.")
+            return
+
+        model_path = Path(model_path)
+        model_config_copy = deepcopy(output_model_config)
+        model_config_copy.config["model_path"] = None
+
+        # upload model config file
+        model_config_bytes = json.dumps(model_config_copy.to_json()).encode()
+        with io.BytesIO(model_config_bytes) as data:
+            self.container_client.upload_blob(f"{output_model_hash}/model_config.json", data=data, overwrite=False)
+
+        # upload model file
+        model_blob = str(Path(output_model_hash) / "model")
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            model_path_file = Path(temp_dir) / "model_path.json"
+            with open(model_path_file, "w") as file:
+                file.write({"model_path": model_path.name})
+            with open(model_path_file, "rb") as data:
+                self.container_client.upload_blob(f"{model_blob}/model_path.json", data=data, overwrite=False)
+
+        if not model_path.is_dir():
+            with open(model_path, "rb") as data:
+                self.container_client.upload_blob(f"{model_blob}/{model_path.name}", data=data, overwrite=False)
+        else:
+            self._upload_dir_to_blob(model_path, f"{model_blob}")
+
+    def _upload_dir_to_blob(self, dir_path, blob_folder_name):
+        for item in dir_path.iterdir():
+            if item.is_dir():
+                self._upload_dir_to_blob(item, f"{blob_folder_name}/{item.name}")
+            else:
+                blob_name = f"{blob_folder_name}/{item.name}"
+                with open(item, "rb") as data:
+                    self.container_client.upload_blob(name=blob_name, data=data, overwrite=False)
+
+
+def check_model_cache(cloud_cache_helper: CloudCacheHelper, input_model_config, pass_search_point, input_model_hash):
+    pass_run_locally = True
+    output_model_config = None
+    logger.info("Cloud model cache is enabled. Check cloud model cache ...")
+
+    output_model_hash = cloud_cache_helper.get_hash_key(input_model_config, pass_search_point, input_model_hash)
+    if cloud_cache_helper.exist_in_cloud_cache(output_model_hash):
+        logger.info("Model is found in cloud cache.")
+        output_model_config = cloud_cache_helper.get_model_config_by_hash_key(output_model_hash)
+        pass_run_locally = False
+    return pass_run_locally, output_model_config
```

Comment:
nit: Unnecessary return value `pass_run_locally`. Just return `None` for `output_model_config`. Two variables create ambiguity where what if `output_model_config` is not None but `pass_run_locally` is still true.

Reply:
Sounds good to me. Will update this.

Reply:
That's engine's logic to decide. I don't see a reason for this function to return the same result in two different logical ways. If engine wants some way to override the result from here it still can do it.

```python
pass_run_locally = not check_model_cache( .... )
# Later down in the logic
if some_other_condition:
  pass_run_locally = False
```

Basically, the caller already has all the information with just returning the `output_model_config`.

```python
output_model_config == (pass_run_locally == False)
(not output_model_config) == (pass_run_locally == True)
```

Reply:
Sounds good to me. Will update this.